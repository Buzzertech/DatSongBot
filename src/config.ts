const config = {
  SOUNDCLOUD_CLIENT_ID: process.env.SOUNDCLOUD_CLIENT_ID,
  SOUNDCLOUD_SEARCH_TAGS: ['Chill', 'Trap'],
  SOUNDCLOUD_SEARCH_LICENSE: 'cc-by',
};

export default config;
