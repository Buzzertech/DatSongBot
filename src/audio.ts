import axios from 'axios';
import { pick, chain, sum } from 'lodash';
import config from './config';
import qs from 'qs';
import { format } from 'date-fns';
import { audioLogger } from './lib/utils';

interface SCUser {
  avatar_url: string;
  id: number;
  kind: 'user';
  permalink_url: string;
  uri: string;
  username: string;
  permalink: string;
  last_modified: Date;
}

export interface Track {
  comment_count: number;
  release: number;
  original_content_size: number;
  track_type: string;
  original_format: string;
  streamable: boolean;
  download_url: string;
  id: number;
  state: 'processing' | 'failed' | 'finished';
  last_modified: Date;
  favoritings_count: number;
  kind: 'track';
  purchase_url: string;
  release_year: number;
  sharing: 'all' | 'private' | 'public';
  attachments_uri: string;
  license:
    | 'no-rights-reserved'
    | 'all-rights-reserved'
    | 'cc-by'
    | 'cc-by-nc'
    | 'cc-by-nd'
    | 'cc-by-sa'
    | 'cc-by-nc-nd'
    | 'cc-by-nc-sa';
  user_id: number;
  user_favorite: boolean;
  waveform_url: string;
  permalink: string;
  permalink_url: string;
  playback_count: number;
  downloadable: boolean;
  created_at: Date;
  description: string;
  title: string;
  duration: number;
  artwork_url: string;
  video_url: string;
  tag_list: string;
  release_month: number;
  genre: string;
  release_day: number;
  reposts_count: number;
  label_name: string;
  commentable: boolean;
  bpm: number;
  policy: string;
  key_signature: string;
  isrc: string;
  uri: string;
  download_count: number;
  likes_count: number;
  purchase_title: string;
  embeddable_by: 'all' | 'me' | 'none';
  monetization_model: string;
  user: SCUser;
  user_playback_count: number;
  stream_url: string;
  label_id: number;
}

export type PickedTrack = Pick<
  Track,
  | 'stream_url'
  | 'download_url'
  | 'user'
  | 'description'
  | 'title'
  | 'purchase_title'
  | 'purchase_url'
  | 'tag_list'
  | 'permalink_url'
  | 'id'
  | 'duration'
>;
interface SCUserWebProfile {
  kind: 'web-profile';
  id: number;
  service: 'personal' | 'youtube' | 'twitter' | 'instagram' | 'facebook';
  title: string;
  url: string;
  username: string;
  created_at: string;
}

export const getTracksFromSoundcloud = async () => {
  try {
    audioLogger('fetching tracks');
    const response = await axios.get<Track[]>(
      `https://api.soundcloud.com/tracks`,
      {
        params: {
          client_id: config.SOUNDCLOUD_CLIENT_ID,
          tags: config.SOUNDCLOUD_SEARCH_TAGS.join(','),
          license: config.SOUNDCLOUD_SEARCH_LICENSE,
          created_at: {
            from: encodeURIComponent(
              format(
                new Date(Date.now() - 60 * 60 * 24 * 1000),
                'YYYY-MM-DD HH:mm:ss'
              )
            ),
            to: encodeURIComponent(format(new Date(), 'YYYY-MM-DD HH:mm:ss')),
          },
        },
        paramsSerializer: params =>
          qs.stringify(params, {
            arrayFormat: 'brackets',
            encode: false,
          }),
      }
    );

    audioLogger(`fetched and got ${response.data.length} responses`);
    const pickedSong = chain(response.data)
      .filter(
        e =>
          e.downloadable ||
          (Boolean(e.stream_url) &&
            e.duration < 60 * 5 * 1000) /* 5 mins or 300000 ms */
      )
      .sort(e => sum([e.playback_count, e.likes_count]))
      .get(0)
      .value();

    audioLogger(
      `Picked Song - ${pickedSong.title} (Soundcloud id - ${pickedSong.id})`
    );

    return pick(pickedSong, [
      'stream_url',
      'download_url',
      'user',
      'description',
      'title',
      'purchase_title',
      'purchase_url',
      'tag_list',
      'permalink_url',
      'id',
      'duration',
    ]);
  } catch (e) {
    audioLogger(`Something went wrong while fetching / picking track`);
    audioLogger(e);
    return Promise.reject(e);
  }
};

const pickUsername = (obj: SCUserWebProfile) =>
  pick(obj, ['username', 'service']);

export const getWebProfileForArtist = async (id: number) => {
  try {
    const { data: webProfiles } = await axios.get<SCUserWebProfile[]>(
      `https://api.soundcloud.com/users/${id}/web-profiles`
    );

    if (webProfiles.length <= 0) {
      return;
    }

    const profilesHash = chain(webProfiles)
      .groupBy('service')
      .mapValues(0)
      .value();

    if (profilesHash.twitter) {
      return pickUsername(profilesHash.twitter);
    } else if (profilesHash.instagram) {
      return pick(profilesHash.instagram);
    } else {
      return;
    }
  } catch (e) {
    return Promise.reject(e);
  }
};
