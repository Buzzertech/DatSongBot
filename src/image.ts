import axios from 'axios';
import config from './config';
import { imageLogger } from './lib/utils';

export interface IUnsplashResponse {
  id: string;
  created_at: string;
  updated_at: string;
  width: number;
  height: number;
  color: string;
  downloads: number;
  likes: number;
  liked_by_user: boolean;
  description: string;
  urls: Record<
    'raw' | 'full' | 'regular' | 'small' | 'thumb' | 'custom',
    string
  >;
  links: Record<'self' | 'html' | 'download' | 'download_location', string>;
  user: {
    id: string;
    updated_at: string;
    username: string;
    name: string;
    portfolio_url: string;
    bio: string;
    location: string;
    total_likes: number;
    total_photos: number;
    total_collections: number;
    instagram_username: string;
    twitter_username: string;
    links: Record<'self' | 'html' | 'photos' | 'likes' | 'portfolio', string>;
  };
}

export const getUnsplashPhoto = async (tags: string) => {
  imageLogger('fetching random photo from unsplash');
  const { data: imageData } = await axios.get<IUnsplashResponse>(
    `https://api.unsplash.com/photos/random`,
    {
      params: {
        client_id: config.UNSPLASH_ACCESS_KEY,
        query: ['nature', ...tags.split(' ')].join(','),
        orientation: 'landscape',
        w: 1920,
        h: 1080,
      },
    }
  );
  imageLogger(`fetched photo from unsplash - ${imageData.links.html}`);
  return imageData;
};
