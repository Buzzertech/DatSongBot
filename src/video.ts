import ffmpeg from 'fluent-ffmpeg';
import { PickedTrack } from './audio';
import chromium from 'chrome-aws-lambda';
import { Browser, LaunchOptions } from 'puppeteer';
import config from './config';
import { IUnsplashResponse } from 'image';
import { imageLogger, videoLogger, durationToSeconds } from './lib/utils';
import installer from '@ffmpeg-installer/ffmpeg';

ffmpeg.setFfmpegPath(installer.path);

let window: Browser;

export const launchPage = async (opts?: LaunchOptions) =>
  (window = await chromium.puppeteer.launch({
    ...opts,
    args: [...(opts && opts.args ? opts.args : []), ...chromium.args],
    defaultViewport: chromium.defaultViewport,
    executablePath: await chromium.executablePath,
  }));

export const closePage = async () => window && (await window.close());

export const prepareSvg = (
  bgUrl: string,
  songName: string,
  artistName: string
) => {
  let textX = '10%';

  if (songName.length > 20) {
    textX = '2%';
  }

  imageLogger(`Preparing svg string for background`);

  return `
		<style>html,body{margin: 0; padding: 0;}</style>
		<link href="https://fonts.googleapis.com/css?family=Poppins&display=swap&text=${songName.toUpperCase()}${artistName.toUpperCase()}" rel="stylesheet">
		<svg viewBox="0 0 1920 1080" lang="en-US" xmlns="http://www.w3.org/2000/svg">
			<defs>
				<linearGradient id="bottomGrad" x1="0%" y1="0%" x2="0%" y2="100%">
					<stop offset="0%" style="stop-color:rgb(255,255,255);stop-opacity:0" />
					<stop offset="100%" style="stop-color:rgb(0,0,0);stop-opacity:.8;" />
				</linearGradient>
			</defs>
			<image href="${bgUrl}" x="0" y="0" width="100%" height="100%" />
			<rect x="0" y="40%" width="100%" height="60%" fill="url(#bottomGrad)"/>
			<text x="${textX}" style="font-family: 'Poppins', arial; font-weight: bold; font-size: 5em;" y="90%" fill="white">${songName.toUpperCase()}</text>
			<text x="${textX}" style="font-family: 'Poppins', arial; font-size: 3em; font-weight: 300;" y="95%" fill="white">${artistName.toUpperCase()}</text>
		</svg>
	`;
};

export const generateImage = async (outputPath: string, content: string) => {
  imageLogger('Launching page in puppeteer');
  const page = await window.newPage();
  imageLogger('Loading svg content into the page');
  await page.setContent(content);
  imageLogger('Loaded svg content into the page');
  await page.setViewport({ width: 1920, height: 1080, isLandscape: true });
  imageLogger(`Processing screenshot`);
  await page.screenshot({
    omitBackground: true,
    fullPage: true,
    path: outputPath,
  });
  imageLogger(`Background image prepared and saved`);
  await window.close();
};

export const processVideo = (
  outputPath: string,
  song: Pick<PickedTrack, 'duration' | 'uri'>,
  image: string
): Promise<void> => {
  videoLogger('Starting to process video');
  videoLogger(`Approx. duration of the video - ${song.duration} ms`);

  const processChain = ffmpeg(image)
    .inputFPS(30)
    //@ts-ignore
    .loop()
    .withSize('1920x1080')
    .input(`${song.uri}/download?client_id=${config.SOUNDCLOUD_CLIENT_ID}`)
    .outputOption('-shortest')
    .videoCodec('libx264')
    .videoBitrate(10000, true)
    .audioCodec('aac')
    .audioBitrate(384)
    .outputOption('-pix_fmt yuv420p')
    .outputFPS(30)
    .videoFilters(
      `fade=in:st=0:d=3,fade=out:st=${song.duration / 1000 - 2}:d=1`
    );

  return new Promise((resolve, reject) => {
    processChain
      .on('start', () => videoLogger(`ffmpeg has begun processing the video`))
      .on('progress', function(progress: { timemark: string }) {
        const duration = durationToSeconds(progress.timemark) * 1000;
        videoLogger(`Current progress - ${(duration / song.duration) * 100}%`);
      })
      .on('end', resolve)
      .on('error', reject)
      .save(outputPath);
  });
};

export const getDescription = (
  songTitle: string,
  song: PickedTrack,
  imageData: IUnsplashResponse
) => `${songTitle}

⭐️ DatSongBot brings you another fresh, new music by ${song.user.username} for you to enjoy!

Listen to this song on Soundcloud:
▶️${song.permalink_url}

Follow ${song.user.username} on Soundcloud:
🔉${song.user.permalink_url}

The background image used in this video is provided by ${imageData.user.name} from Unsplash:
🔗Follow ${imageData.user.name} on Unsplash - ${imageData.user.links.html}
📂Download this background - ${imageData.links.html}

🎵 DatSongBot is a bot built by Buzzertech (https://buzzertech.com) which picks a new, trending song from soundcloud and uploads it to YouTube. This is an experimental tool. With that being said, be sure to subscribe to DatSongBot on YouTube and turn on notifications 'cause we post new music daily on this channel!

❌ DatSongBot doesn't owns this music nor the image used in this video. Just for entertainment purposes only!

Cheers 🎵
  `;
