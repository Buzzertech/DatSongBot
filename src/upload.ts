import { google, youtube_v3 } from 'googleapis';
import config from './config';
import { videoLogger } from './lib/utils';
import { getDescription } from './video';
import { stat, createReadStream } from 'fs-extra';
import { PickedTrack } from './audio';
import { IUnsplashResponse } from './image';
import { addDays } from 'date-fns';

export const connectToYoutube = async () => {
  const oauthclient = new google.auth.OAuth2({
    clientId: config.YOUTUBE_CLIENT_ID,
    clientSecret: config.YOUTUBE_CLIENT_SECRET,
  });

  oauthclient.setCredentials({
    refresh_token: config.YOUTUBE_REFRESH_TOKEN,
  });

  const { token: access_token } = await oauthclient.getAccessToken();

  oauthclient.setCredentials({
    access_token,
  });

  return google.youtube({ version: 'v3', auth: oauthclient });
};

export const uploadVideo = async (
  videoPath: string,
  song: PickedTrack,
  imageData: IUnsplashResponse,
  youtubeClient: youtube_v3.Youtube
) => {
  const { size: totalVideoByteSize } = await stat(videoPath);

  const songTitle =
    song.title.replace(/(")|(')|(\.)/g, '').trim() + ` | ${song.user.username}`;

  videoLogger(`Preparing the video description`);

  const description = getDescription(songTitle, song, imageData);

  return youtubeClient.videos.insert(
    {
      part: 'snippet, status',
      requestBody: {
        snippet: {
          title: songTitle,
          description,
          categoryId: '10',
          tags: [...song.tag_list.split(' '), 'DatSongBot', 'ZeonBot', 'Music'],
          defaultLanguage: 'en',
        },
        status: {
          embeddable: false,
          privacyStatus: 'private',
          license: 'youtube',
          publishAt: addDays(new Date(), 4).toISOString(),
        },
      },
      media: {
        body: createReadStream(videoPath),
      },
    },
    {
      onUploadProgress: progress => {
        videoLogger(
          `Video upload progress - ${new Number(
            (progress.bytesRead / totalVideoByteSize) * 100
          ).toFixed(2)}%`
        );
      },
    }
  );
};
