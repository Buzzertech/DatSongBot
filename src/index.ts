import {
  generateImage,
  prepareSvg,
  launchPage,
  closePage,
  processVideo,
} from './video';
import { getTracksFromSoundcloud } from './audio';
import { getUnsplashPhoto } from './image';
import { videoLogger } from './lib/utils';
import { uploadVideo, connectToYoutube } from './upload';
import { init as initSentry } from '@sentry/node';
import config from './config';
const IMAGE_OUTPUT = '/tmp/out.png';
const VIDEO_OUTPUT = '/tmp/out.mp4';

if (process.env.NODE_ENV === 'production') {
  global.__rootdir__ = __dirname || process.cwd();

  initSentry({
    dsn: config.SENTRY_DSN,
  });
}

type TReturnValue = () => Promise<void>;

export const main: TReturnValue = async () => {
  try {
    const [youtubeClient, song] = await Promise.all([
      connectToYoutube(),
      getTracksFromSoundcloud(),
      launchPage(),
    ]);
    const image = await getUnsplashPhoto(song.tag_list);
    const svgContent = prepareSvg(
      image.urls.custom,
      song.title.replace(/(")|(')|(\.)/g, ''),
      song.user.username
    );
    await generateImage(IMAGE_OUTPUT, svgContent);
    await processVideo(VIDEO_OUTPUT, song, IMAGE_OUTPUT);
    const response = await uploadVideo(
      VIDEO_OUTPUT,
      song,
      image,
      youtubeClient
    );

    videoLogger(`Video has been uploaded!`);
    videoLogger(`Youtube video id - ${response.data.id}`);

    closePage();
  } catch (e) {
    await closePage();
    console.error(e);
    return Promise.reject(e);
  }
};
