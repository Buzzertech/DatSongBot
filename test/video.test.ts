import { launchPage, closePage, prepareSvg, generateImage } from '../src/video';
import { puppeteer } from 'chrome-aws-lambda';
import { LaunchOptions } from 'puppeteer';
import fs from 'fs-extra';
import {
  toMatchImageSnapshot,
  configureToMatchImageSnapshot,
} from 'jest-image-snapshot';

configureToMatchImageSnapshot({
  customDiffConfig: {
    threshold: 0.08,
  },
});

expect.extend({ toMatchImageSnapshot });

const PUPPETEER_OPTS: LaunchOptions = {
  args: ['-disable-sandbox', 'â€“disable-setuid-sandbox'],
};

describe('video', () => {
  afterAll(async () => await closePage());
  describe('#launchPage', () => {
    let launchSpy: jest.SpyInstance;

    afterEach(async () => {
      await closePage();
    });

    beforeAll(() => {
      launchSpy = jest.spyOn(puppeteer, 'launch');
    });

    afterAll(() => jest.restoreAllMocks());

    it('create a new browser instance', async () => {
      await launchPage(PUPPETEER_OPTS);
      expect(launchSpy).toHaveBeenCalled();
    });
  });

  describe('#closePage', () => {
    let closePageSpy: jest.SpyInstance;

    beforeAll(async () => {
      const window = await launchPage(PUPPETEER_OPTS);
      closePageSpy = jest.spyOn(window, 'close');
    });

    afterAll(() => jest.restoreAllMocks());

    it('will close the browser', async () => {
      await closePage();
      expect(closePageSpy).toHaveBeenCalled();
    });
  });

  describe('#prepareSvg', () => {
    it('execute the template and return a svg string', () => {
      const svgString = prepareSvg('google.com', 'Sandstorm', 'Darude');
      expect(svgString).toMatchSnapshot();
    });
  });

  describe('#generateImage', () => {
    let svgStr: string;

    beforeAll(async () => {
      await launchPage(PUPPETEER_OPTS);
      svgStr = prepareSvg(
        'https://placehold.it/1920x1080',
        'Sandstorm',
        'Darude'
      );
    });

    afterAll(async () => await closePage());

    it('generate a png image', async () => {
      await generateImage('/tmp/out.png', svgStr);
      const image = await fs.readFile('/tmp/out.png');
      jest.retryTimes(3);
      expect(image).toMatchImageSnapshot({
        customSnapshotIdentifier: 'sandstorm',
      });
    });
  });

  describe('#processVideo', () => {
    it('will create a video file', () => {});
  });
});
