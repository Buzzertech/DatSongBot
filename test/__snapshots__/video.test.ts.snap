// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`video #getDescription generates a description string 1`] = `
"Sandstorm

⭐️ DatSongBot brings you another fresh, new music by Matthew186f for you to enjoy!

Listen to this song on Soundcloud:
▶️http://soundcloud.com/user-525309592/wave-he

Follow Matthew186f on Soundcloud:
🔉http://soundcloud.com/user-525309592

The background image used in this video is provided by Chris Benson from Unsplash:
🔗Follow Chris Benson on Unsplash - https://unsplash.com/@lordmaui
📂Download this background - https://unsplash.com/photos/nKEARsgmrqc

🎵 DatSongBot is a bot built by Buzzertech (https://buzzertech.com) which picks a new, trending song from soundcloud and uploads it to YouTube. This is an experimental tool. With that being said, be sure to subscribe to DatSongBot on YouTube and turn on notifications 'cause we post new music daily on this channel!

❌ DatSongBot doesn't owns this music nor the image used in this video. Just for entertainment purposes only!

Cheers 🎵
  "
`;

exports[`video #prepareSvg execute the template and return a svg string 1`] = `
"
		<style>html,body{margin: 0; padding: 0;}</style>
		<link href=\\"https://fonts.googleapis.com/css?family=Poppins&display=swap&text=SANDSTORMDARUDE\\" rel=\\"stylesheet\\">
		<svg viewBox=\\"0 0 1920 1080\\" lang=\\"en-US\\" xmlns=\\"http://www.w3.org/2000/svg\\">
			<defs>
				<linearGradient id=\\"bottomGrad\\" x1=\\"0%\\" y1=\\"0%\\" x2=\\"0%\\" y2=\\"100%\\">
					<stop offset=\\"0%\\" style=\\"stop-color:rgb(255,255,255);stop-opacity:0\\" />
					<stop offset=\\"100%\\" style=\\"stop-color:rgb(0,0,0);stop-opacity:.8;\\" />
				</linearGradient>
			</defs>
			<image href=\\"google.com\\" x=\\"0\\" y=\\"0\\" width=\\"100%\\" height=\\"100%\\" />
			<rect x=\\"0\\" y=\\"40%\\" width=\\"100%\\" height=\\"60%\\" fill=\\"url(#bottomGrad)\\"/>
			<text x=\\"10%\\" style=\\"font-family: 'Poppins', arial; font-weight: bold; font-size: 5em;\\" y=\\"90%\\" fill=\\"white\\">SANDSTORM</text>
			<text x=\\"10%\\" style=\\"font-family: 'Poppins', arial; font-size: 3em; font-weight: 300;\\" y=\\"95%\\" fill=\\"white\\">DARUDE</text>
		</svg>
	"
`;
